scipy/sparse/csgraph/_flow.pyx

```diff
@@ -0,0 +1,395 @@
+import numpy as np
+
+from scipy.sparse import csr_matrix, isspmatrix_csr
+
+cimport cython
+cimport numpy as np
+
+
+class MaximumFlowResult:
+    """Represents the result of a maximum flow calculation.
+
+    Attributes
+    ----------
+    flow_value : int
+        The value of the maximum flow.
+    residual : csr_matrix
+        The residual graph with respect to the maximum flow.
+    """
+
+    def __init__(self, flow_value, residual):
+        self.flow_value = flow_value
+        self.residual = residual
+
+
+def maximum_flow(csgraph, source, sink):
+    r"""
+    maximum_flow(csgraph, source, sink)
+                  
+    Maximize the flow between two vertices in a graph.
+
+    .. versionadded:: 1.4.0
+
+    Parameters
+    ----------
+    csgraph : csr_matrix
+        The square matrix representing a directed graph whose (i, j)'th entry
+        is an integer representing the capacity of the edge between
+        vertices i and j.
+    source : int
+        The source vertex from which the flow flows.
+    sink : int
+        The sink vertex to which the flow flows.
+
+    Returns
+    -------
+    res : MaximumFlowResult
```

Comment:
This is a little unusual, but it does allow for extending the returns in the future. So I think it's fine. The one thing I'm wondering about is whether to put this class in the `csgraph` namespace - that pattern then adds two things to a namespace each time a single function is added. I'm inclined to not do that. In that case, the line below should use double backticks, because there won't be a doc page for `MaximumFlowResult` to link to.

Reply:
Right, this might well be looking too far into the future. A place where having `MaximumFlowResult` would pay off is the following: Assume that the feature turns out to be popular; then a way to improve performance even further would be to expand the set of solver methods (low-hanging fruit would be [push–relabel](https://en.wikipedia.org/wiki/Push%E2%80%93relabel_maximum_flow_algorithm)); doing so, the collection of outputs from the general purpose `maximum_flow` entry point would depend on the solver, and collecting them in `MaximumFlowResult` would provide flexibility without the risk of breaking API changes.

I guess the class wouldn't be particularly useful outside of `maximum_flow` itself, so putting it in `csgraph` won't buy you much either? Perhaps other than consistency?

I'm not too familiar with Sphinx syntax; it [looks like](https://stackoverflow.com/a/22257273/5085211) the use of double backticks, but would I be right in understanding that single backticks are used whenever additional parsing, and double backticks are used everywhere else?